### PR TITLE
websocket client adapter

### PR DESCRIPTION
**Pull request by wrieg123**
Added a Websocket Client Adapter (w/ and w/o TLS support).

Further, there is an output adapter for updating headers on reconnects.

### CODE REVIEW

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:14]: Can we potentially avoid a copy here with `std::move( properties )`?

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:75]: We can accept the message as a `const std::string &` as long as `client.send` is copying into a buffer

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:107]: Is the "Good bye" standard or put in for testing?

File: cpp/csp/adapters/websockets/ClientOutputAdapter.h
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/ClientOutputAdapter.h:28]: As a general rule I think it's better to include headers where they are used rather than to let them trickle in from other includes, so let's include the relevant `websocketpp` header here. Two reasons for this:

1) If dependencies change in another file it won't affect this file
2) Easier to find where the code is coming from if the include path is present

File: cpp/csp/adapters/websockets/ClientOutputAdapter.cpp
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/ClientOutputAdapter.cpp:2]: Unused iostream include, maybe for prior debugging?

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp:2]: Unused iostream include, same as my other comment

File: cpp/csp/adapters/websockets/ClientAdapterManager.h
Comment (AdamGlustein) [inline cpp/csp/adapters/websockets/ClientAdapterManager.h:71]: Isn't this redundant since you can use `if (m_thread)` instead (it will be nullptr before you launch the thread)?

**Review by AdamGlustein — DISMISSED** [review]
Made a first pass through just focusing on the `csp` related code as I'm not familiar with how websockets work.

File: cpp/csp/adapters/websockets/ClientInputAdapter.cpp
Comment (timkpaine) [reply cpp/csp/adapters/websockets/ClientInputAdapter.cpp:37]: Don't, the autoformatter has not yet been applied to the cpp codebase

File: csp/adapters/websocket_client.py
Comment (timkpaine) [reply csp/adapters/websocket_client.py:60]: Yes

File: conda/dev-environment-linux.yml
Comment (timkpaine) [inline conda/dev-environment-linux.yml:7]: Please use the new library, either `libboost-headers` or `libboost`

**Review by timkpaine — CHANGES_REQUESTED** [review]

File: conda/dev-environment-linux.yml
Comment (timkpaine) [reply conda/dev-environment-linux.yml:50]: It's going to be some time before there is a `csp-extensions` repo in any form, and certainly for C++ extensions where there may be ABI boundaries. I don't think we'd want to block this PR on waiting (since there are active uses of a websocket client).

File: csp/adapters/websocket_client.py
Comment (wrieg123) [reply csp/adapters/websocket_client.py:105]: Yes, it supports the message mappers so you could get a struct

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (wrieg123) [reply cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:107]: I just chose something, I guess it could be anything

File: csp/adapters/websocket_client.py
Comment (wrieg123) [reply csp/adapters/websocket_client.py:32]: added

File: examples/3_using_adapters/websocket_client.py
Comment (wrieg123) [reply examples/3_using_adapters/websocket_client.py]: updated example to do both

File: csp/adapters/websocket_client.py
Comment (AdamGlustein) [reply csp/adapters/websocket_client.py]: So why is use_tls an option if the URi starts with 'wss'?

Comment (wrieg123) [reply csp/adapters/websocket_client.py]: at the time, I didn't want to make any inference. I've removed the kwarg and updated it to be startswith('wss')

File: csp/adapters/websocket_client.py
Comment (wrieg123) [reply csp/adapters/websocket_client.py]: i've updated the naming to be reconnect_interval; ill look into writing it as a timedelta

File: CMakeLists.txt
Comment (timkpaine) [inline CMakeLists.txt:74]: Please enable the corresponding flag in `setup.py` (it was merged yesterday, apologies)

**Review by timkpaine — CHANGES_REQUESTED** [review]

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (timkpaine) [reply cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:107]: the symphony adapter does something similar and I kind of dislike it. Is it possible to configure this? considering its on websocket close, it doesnt really seem like you need to send anything, maybe an empty string is sufficient?

File: csp/impl/wiring/numba_node.py
Comment (timkpaine) [inline csp/impl/wiring/numba_node.py:70]: can you double check that you're using `ruff>0.3,<0.4`, i just did an autofix yesterday and i dont think this change should happen.

Comment (wrieg123) [reply csp/impl/wiring/numba_node.py:70]: ruff 0.3.4

File: cpp/csp/adapters/websockets/CMakeLists.txt
Comment (timkpaine) [inline cpp/csp/adapters/websockets/CMakeLists.txt:22]: Is this strictly required?

**Review by timkpaine — DISMISSED** [review]
I'm comfortable with the `vcpkg`, `cmake`, `conda`, and `setup.py` changes. The remaining will require other people to review

File: cpp/csp/adapters/websockets/CMakeLists.txt
Comment (wrieg123) [reply cpp/csp/adapters/websockets/CMakeLists.txt:22]: I guess it doesn't explicitly need to search for the static libs

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (wrieg123) [reply cpp/csp/adapters/websockets/WebsocketEndpoint.cpp]: I adjusted this to be an empty string. We could always add it as a property (on_close_message or something similar) if there is a use case.

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:78]: should we check EC here / report an error on the status adapter?

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:94]: check should go after the get_connection call, I assume if ec is set then con can be null and the loop would crash

File: cpp/csp/adapters/websockets/WebsocketEndpoint.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/WebsocketEndpoint.cpp:129]: same comments below
- lambda callbacks
- check ec on send
- check ec after get_connection call

File: cpp/csp/adapters/websockets/ClientAdapterManager.h
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientAdapterManager.h:49]: name should have websocket in it somewhere

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientAdapterManager.cpp:44]: would it make more sense to check m_inputAdapter != null before setting the CB instead?

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientAdapterManager.cpp:75]: can you explain what shouldRun signifies vs active?
is it possible for m_active to become true again  in the next iteration here?

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientAdapterManager.cpp:79]: drop std chrono, just call csp::sleep( ...get<TimeDelta> )

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientAdapterManager.cpp:88]: you probably want to do this after you join the thread, otherwise you can race / crash

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.h
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.h:13]: may want to call this ClientHeaderUpdateOutputAdapter for clarity

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp:15]: access by reference (&) to avoid an expensive copy

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp:17]: if you want it to be less hacky ( with slightly more work ) you can define a csp.Struct class with the key : str and value : str fields, and make headers an array of that.
You would then have to add autogen to generate the c++ headers to access to the data ( or, you dont, and you can look up the fields dynamically too )

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp:22]: do you want to support clearing a header too?
if you do make it a struct you can take an unset value as a clear

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.h
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.h:23]: same comment on the rest of the name() methods, add websocket in the name somewhere

File: cpp/csp/adapters/websockets/ClientOutputAdapter.h
Comment (robambalu) [inline cpp/csp/adapters/websockets/ClientOutputAdapter.h:30]: unused var

File: cpp/csp/python/adapters/websocketsadapterimpl.cpp
Comment (robambalu) [inline cpp/csp/python/adapters/websocketsadapterimpl.cpp:47]: same comment, any reference to this adapter that would go up to the user should have websocket in the name

File: cpp/csp/adapters/websockets/CMakeLists.txt
Comment (wrieg123) [reply cpp/csp/adapters/websockets/CMakeLists.txt:1]: the reason I named it client was because of the "websockets" table adapter that already exists

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (wrieg123) [reply cpp/csp/adapters/websockets/ClientAdapterManager.cpp:38]: I think @timkpaine is going to autoformat all the c++ code

File: cpp/csp/adapters/websockets/CMakeLists.txt
Comment (robambalu) [reply cpp/csp/adapters/websockets/CMakeLists.txt:1]: I didnt realize there was already a csp.adapters.websockets module, in which case we should discuss it.
I wouldnt want to have websockets and websocket_client adapters, that'll just be confusing.
What does the current websocket adapter actually do, and is it something your adapter can replace?

Comment (wrieg123) [reply cpp/csp/adapters/websockets/CMakeLists.txt:1]: the current one is broadcasting what looks to be tables (maybe for perspective?): https://github.com/Point72/csp/blob/main/csp/adapters/websocket.py

Comment (timkpaine) [reply cpp/csp/adapters/websockets/CMakeLists.txt:1]: non-perspective table-like things. Its on the chopping block, but its a websocket-server not a websocket client.

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (wrieg123) [reply cpp/csp/adapters/websockets/ClientAdapterManager.cpp:88]: not sure what you mean, close() terminates the ws cnx to the server, therefore the loop will evaluate m_shouldRun, see that it should not try to reconnect, then join later on

File: cpp/csp/adapters/websockets/CMakeLists.txt
Comment (wrieg123) [reply cpp/csp/adapters/websockets/CMakeLists.txt]: moved into websocket.py

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp
Comment (wrieg123) [reply cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp]: I'm not sure dictionaries have a clear in the cpp impl. https://github.com/Point72/csp/blob/main/cpp/csp/engine/Dictionary.h#L25

**Review by timkpaine — DISMISSED** [review]
I forgot the `NOTICE` also has to be regenerated

File: cpp/csp/adapters/websocket/ClientAdapterManager.cpp
Comment (robambalu) [inline cpp/csp/adapters/websocket/ClientAdapterManager.cpp:9]: are these used anymore?

File: cpp/csp/adapters/websocket/WebsocketEndpoint.h
Comment (robambalu) [inline cpp/csp/adapters/websocket/WebsocketEndpoint.h:48]: more typedefs.  replace all typedefs with using

File: cpp/csp/adapters/websocket/WebsocketEndpoint.h
Comment (robambalu) [inline cpp/csp/adapters/websocket/WebsocketEndpoint.h:23]: previous comment wasnt addressed, should have a virtual dtor

File: cpp/csp/adapters/websockets/ClientAdapterManager.cpp
Comment (robambalu) [reply cpp/csp/adapters/websockets/ClientAdapterManager.cpp]: I see, run() is blocking and close would get it to unblock / exit the thread

File: cpp/csp/adapters/websocket/WebsocketEndpoint.h
Comment (wrieg123) [reply cpp/csp/adapters/websocket/WebsocketEndpoint.h:48]: thanks, forgot about those

File: cpp/csp/adapters/websocket/ClientAdapterManager.cpp
Comment (wrieg123) [reply cpp/csp/adapters/websocket/ClientAdapterManager.cpp:9]: no, will remove

File: cpp/csp/adapters/websocket/WebsocketEndpoint.h
Comment (wrieg123) [reply cpp/csp/adapters/websocket/WebsocketEndpoint.h:23]: for got to commit, will change

File: cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp
Comment (robambalu) [reply cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp]: They do not, thats why I said "if you do make it a struct"

Comment (wrieg123) [reply cpp/csp/adapters/websockets/ClientHeaderUpdateAdapter.cpp]: I think for now, I'm going to hold off on clearing a header or making the headers a struct themselves. If there is interest down the line it is something to look into supporting.

**Review by robambalu — APPROVED** [review]

File: conda/dev-environment-unix.yml
Comment (timkpaine) [inline conda/dev-environment-unix.yml:21]: @wrieg123 alphabetize!

**Pull request merged**
